### ISSUE TITLE
404 not found - http://apt.llvm.org/focal/

### ISSUE THREAD

**Issue by yeongrokgim** [llvm/llvm-project#64097]
Dear maintainers, currently `focal` apt repository is unavailable.

## Suggested labels

- [packaging](https://github.com/llvm/llvm-project/labels/packaging)
- [website](https://github.com/llvm/llvm-project/labels/website)
- [platform:linux](https://github.com/llvm/llvm-project/labels/platform%3Alinux)

## Potentially related issues

- https://github.com/llvm/llvm-project/issues/63998
  - https://github.com/llvm/llvm-project/issues/64003

## Details

It looks like `focal` apt repository is having ongoing maintenance (or an incident).

- `focal` is failing (same result for `https://apt.llvm.org/focal/dists/llvm-toolchain-focal-15/Release`)
  
  ```bash
  $ curl https://apt.llvm.org/focal/ -H "Accept: text/html" -sv | gunzip
  *   Trying 199.232.198.49:443...
  * TCP_NODELAY set
  * Connected to apt.llvm.org (199.232.198.49) port 443 (#0)
  * ALPN, offering h2
  * ALPN, offering http/1.1
  * successfully set certificate verify locations:
  *   CAfile: /etc/ssl/certs/ca-certificates.crt
    CApath: /etc/ssl/certs
  } [5 bytes data]
  * TLSv1.3 (OUT), TLS handshake, Client hello (1):
  } [512 bytes data]
  * TLSv1.3 (IN), TLS handshake, Server hello (2):
  { [106 bytes data]
  * TLSv1.2 (IN), TLS handshake, Certificate (11):
  { [2809 bytes data]
  * TLSv1.2 (IN), TLS handshake, Server key exchange (12):
  { [300 bytes data]
  * TLSv1.2 (IN), TLS handshake, Server finished (14):
  { [4 bytes data]
  * TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
  } [37 bytes data]
  * TLSv1.2 (OUT), TLS change cipher, Change cipher spec (1):
  } [1 bytes data]
  * TLSv1.2 (OUT), TLS handshake, Finished (20):
  } [16 bytes data]
  * TLSv1.2 (IN), TLS handshake, Finished (20):
  { [16 bytes data]
  * SSL connection using TLSv1.2 / ECDHE-RSA-AES128-GCM-SHA256
  * ALPN, server accepted to use h2
  * Server certificate:
  *  subject: CN=*.llvm.org
  *  start date: May  6 21:55:00 2023 GMT
  *  expire date: Jun  6 21:54:59 2024 GMT
  *  subjectAltName: host "apt.llvm.org" matched cert's "*.llvm.org"
  *  issuer: C=BE; O=GlobalSign nv-sa; CN=GlobalSign Atlas R3 DV TLS CA 2023 Q2
  *  SSL certificate verify ok.
  * Using HTTP2, server supports multi-use
  * Connection state changed (HTTP/2 confirmed)
  * Copying HTTP/2 data in stream buffer to connection buffer after upgrade: len=0
  } [5 bytes data]
  * Using Stream ID: 1 (easy handle 0xaaaad0a600b0)
  } [5 bytes data]
  > GET /focal/ HTTP/2
  > Host: apt.llvm.org
  > user-agent: curl/7.68.0
  > accept: text/html
  > 
  { [5 bytes data]
  * Connection state changed (MAX_CONCURRENT_STREAMS == 100)!
  } [5 bytes data]
  < HTTP/2 404 
  < server: nginx/1.4.6 (Ubuntu)
  < content-type: text/html
  < content-encoding: gzip
  < via: 1.1 varnish, 1.1 varnish
  < accept-ranges: bytes
  < date: Tue, 25 Jul 2023 10:49:28 GMT
  < age: 207
  < x-served-by: cache-bur-kbur8200136-BUR, cache-lax10622-LGB
  < x-cache: HIT, HIT
  < x-cache-hits: 4, 1
  < x-timer: S1690282169.900652,VS0,VE3
  < strict-transport-security: max-age=300
  < content-length: 197
  < 
  { [197 bytes data]
  * Connection #0 to host apt.llvm.org left intact
  <html>
  <head><title>404 Not Found</title></head>
  <body bgcolor="white">
  <center><h1>404 Not Found</h1></center>
  <hr><center>nginx/1.4.6 (Ubuntu)</center>
  </body>
  </html>
  <!-- a padding to disable MSIE and Chrome friendly error page -->
  <!-- a padding to disable MSIE and Chrome friendly error page -->
  <!-- a padding to disable MSIE and Chrome friendly error page -->
  <!-- a padding to disable MSIE and Chrome friendly error page -->
  <!-- a padding to disable MSIE and Chrome friendly error page -->
  <!-- a padding to disable MSIE and Chrome friendly error page -->
  ```

- `bionic` or `jammy` is working properly(same result for `https://apt.llvm.org/{bionic,jammy}/dists/llvm-toolchain-{bionic,jammy}-15/Release`)
  
  ```bash
  $ curl https://apt.llvm.org/bionic/ -H "Accept: text/html" -s 
  <html>
  <head><title>Index of /bionic/</title></head>
  <body bgcolor="white">
  <h1>Index of /bionic/</h1><hr><pre><a href="../">../</a>
  <a href="conf/">conf/</a>                                              16-May-2023 19:16                   -
  <a href="db/">db/</a>                                                24-Jul-2023 11:23                   -
  <a href="dists/">dists/</a>                                             25-Jan-2023 19:21                   -
  <a href="pool/">pool/</a>                                              11-Jan-2018 20:13                   -
  </pre><hr></body>
  </html>
  
  $ curl https://apt.llvm.org/jammy/ -H "Accept: text/html" -s
  <html>
  <head><title>Index of /jammy/</title></head>
  <body bgcolor="white">
  <h1>Index of /jammy/</h1><hr><pre><a href="../">../</a>
  <a href="conf/">conf/</a>                                              30-Jan-2023 22:30                   -
  <a href="db/">db/</a>                                                24-Jul-2023 08:41                   -
  <a href="dists/">dists/</a>                                             30-Jan-2023 23:05                   -
  <a href="pool/">pool/</a>                                              16-Feb-2022 13:14                   -
  </pre><hr></body>
  </html>
  ```

**Comment by vient** [comment]
Looking at https://apt.llvm.org, from all Ubuntu distros only `focal` is affected (based on missing "Last update" info)

![image](https://github.com/llvm/llvm-project/assets/7602242/63da9de6-4466-4c40-bc23-dce41b65735b)

**Comment by WheelerCC** [comment]
Seeing the same

**Comment by sylvestre** [comment]
working on it

**Comment by sylvestre** [comment]
Fixed. I added a check to prevent the orignal directory to be removed

For the context, to avoid CDN mismatch, the algorithm is something like:

1. Copy the current repo to a new directory to be updated
2. Sync the data from the build server to the apt repo
3. Delete the current repo (by renaming it)
4. Move the new repo to the actual dir
5.Delete the old repo

it was failing between 3 & 4. I merged 3 and 4 into a single command to prevent this from happening
